src/Components/Components/test/Routing/RouterTest.cs

```diff
@@ -2,103 +2,203 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Reflection;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.AspNetCore.Components;
 using Microsoft.AspNetCore.Components.Routing;
 using Microsoft.AspNetCore.Components.Test.Helpers;
-using Microsoft.Extensions.DependencyModel;
+using Microsoft.Extensions.Logging.Abstractions;
 using Moq;
 using Xunit;
 
 namespace Microsoft.AspNetCore.Components.Test.Routing
 {
     public class RouterTest
     {
+        private readonly Mock<Router> _router;
+        private readonly TestRenderer _renderer;
+
+        public RouterTest()
+        {
+            _router = CreateMockRouter();
```

Comment:
Totally up to you, but if you were to use [this pattern](https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/test/OwningComponentBaseTest.cs#L19-L26) to get `_router` from the `_renderer`, then it wouldn't be necessary to make the `NavigationManager`/`LoggerFactory` properties artificially `internal`. It can still use mock service instances.

Reply:
Thanks for this reference! I've updated the test file to use this pattern. I ended up being able to remove all the mocks together. Of course, I had to do a little bit more to make sure that the `Refresh` method worked. Mostly, this was just:

- bringing back the route components
- adding some necessary parameters to make sure the refresh method still ran successfully

Also, I ended up needing to invoke the methods inside `_renderer.Dispatcher.InvokeAsync` to ensure that they were invoked in the renderer context and not the AsyncTestContext.